### DISCUSSION TITLE
mpremote exec: semicolon for "one-liners" works. Will it alway work?

### DISCUSSION THREAD

**Discussion by DavesCodeMusings** [micropython/micropython#11576] [General]
I've been looking for a way to run multiple MicroPython statements on one line using `mpremote exec`. It seems the semicolon works similar to the way one would expect in a BASH shell.

```
PS C:\> py -m mpremote exec "from os import listdir; print(listdir())"
['boot.py', 'device_specs.py']
```

This is fabulous, but also unexpected since Python doesn't end lines with semicolons like other languages. `mpremote exec` doesn't seem to mind the spaces trailing the semicolon either. Serendipity!

But, didn't see this feature mentioned in any documentation.

I'm happy it works, but is it an intentional feature? If I start writing exec strings like BASH one-liners will it break in some future release of mpremote?

**Comment by jimmo** [comment]
This is standard Python (and therefore MicroPython). Semicolons are optional as a statement separator in Python.

**Reply by jimmo** [reply]
And just to be clear w.r.t. your question, none of this is being handled or interpreted by myremote.
mpremore just passes the exec command directly to the REPL.

**Reply by DavesCodeMusings** [reply]
Thank you for the explanation. I was always under the impression that semicolons were taboo in Python. I learned something new today.

**Reply by peterhinch** [reply]
They are allowed by the language. Some disapprove of them on style/readability grounds, although PEP8 doesn't seem to have a view.

**Comment by DavesCodeMusings** [comment]
Here's a couple one-liners I came up with in the end...

Show the files in a directory:
```
from os import listdir, stat ; print([entry for entry in listdir() if stat(entry)[0] & 0x8000 != 0])
```

Show the subdirs in a directory:
```
from os import listdir, stat ; print([entry for entry in listdir() if stat(entry)[0] & 0x4000 != 0])
```

My goal was to use `mpremote exec` to grab a list of files from the microcontroller and then offer a selection list in a Visual Studio Code extension. The output from the one-liners is close enough to JSON that it can be parsed by the VSCode extension without much trouble. Just wanted to share in case anyone else is thinking of doing something similar.

**Reply by peterhinch** [reply]
You might be interested in [this](https://github.com/Josverl/mpremote_config). It enables `mpremote` to be extended with your own custom scripts. For example I can run
```bash
$ mpremote wifi_connect
```
and `mpremote` will copy `wifi_connect.py` to the target's RAM and execute it.

**Reply by DavesCodeMusings** [reply]
That looks pretty cool.
